
# MTSG

This repository contains the implementation of the **MTSG** method for identifying associations between splicing events and the genetics of complex diseases.

---

## Steps to Run

### 1. Download the Data

Download `data.tar` from [this link](https://www.dropbox.com/scl/fo/vly3z0mxawa5x0v9k4elh/ADUtksnvW0Iyo71twsQDNGc?rlkey=vi69osfby7ipnjru9wva1tte3&dl=0).
The archive contains the following files:

* **gene.500k.id** – Coordinates of the 500 kb upstream and downstream regions for each gene.
* **hg38.vcf** – rsIDs for each SNP in the 1000 Genomes Project.
* **GTEx\_Analysis\_2017-06-05\_v8\_WholeGenomeSeq\_838Indiv\_Analysis\_Freeze.SHAPEIT2\_phased.vcf.gz** – Genotype data for all GTEx individuals, used for generating the X matrix.
* **samples.filtered.used** – The sample list used for running the MSG methods. This needs to be generated by users for each tissue. You may include all samples from the genotype file; filtering will be done automatically during execution.
* **LDREF** – Linkage disequilibrium reference files for each chromosome.
* **all\_splicing\_filled\_expand** – Splicing expression data in GTEx, used for generating the Y matrix.
* **sumstats** – GWAS summary statistics for a specific disease. Users need to format their own file in the required format.

---

### 2. Clone the Code

Clone this repository to your local machine:

* **scripts/** – Contains all scripts required for the MTSG method.
* **runRmd.R** – Runs `.Rmd` files from the command line with user-specified options.
* **perGene.process.x.rmd** – Generates the X matrix for each gene.
* **perGene.process.y.rmd** – Generates the Y matrix for each gene.
* **generate\_db\_and\_cov.R** – Generates the `.cov` file for each gene.
* **gtex\_comp\_MSG\_ACAT\_GBJ\_120522.R** – Runs the MSG method.
* **impute\_missing\_0120.R** – Imputes missing splicing expression values.
* **sda\_static\_linux** – Binary for tensor decomposition.
* **others/** – Additional required scripts.

---

### 3. Prepare Data and Code

Place the data and code in the same directory.
If using different directories, specify the absolute paths when running commands.

Example: Run for gene `ENSG00000000457`

---

## Step-by-Step Instructions

### Using Conda to Set Up the Environment

```bash
# 1. Install Conda (Mambaforge variant recommended)
ENV=/path_to_conda/miniconda3

mkdir download/
cd download/
wget https://github.com/conda-forge/miniforge/releases/latest/download/Mambaforge-Linux-x86_64.sh
mkdir -p $ENV
bash Mambaforge-Linux-x86_64.sh -b -p $ENV

export PATH=$PATH:$ENV/bin

# 2. Install R, Perl, cpanm, GNU Parallel, and required R packages
source $ENV/bin/activate
mamba install -y -c conda-forge r-base perl perl-app-cpanminus parallel r-data.table r-optparse r-magrittr r-reshape2 r-pma r-mvtnorm r-rcpp r-devtools r-stringi r-rcppeigen r-gbj r-glmnet r-this.path r-dplyr r-devtools

# 3. Link compilers
parallel echo ln -s $ENV/bin/x86_64-conda_cos6-linux-gnu-cc $ENV/bin/{=s/.*-//=} ::: $ENV/bin/x86_64-conda_cos6-linux-gnu-cc $ENV/bin/x86_64-conda_cos6-linux-gnu-gcc $ENV/bin/x86_64-conda_cos6-linux-gnu-g++ | bash

# 4. Install Perl modules
mamba install -y -c bioconda perl-dbi perl-ipc-run perl-mce perl-module-build perl-string-random perl-mce-shared
cpanm Switch List::Uniq
cpanm Array/Utils.pm MCE/Hobo.pm
cpanm --force Env/Modify.pm
wget https://github.com/crotoc/Bundle-Wrapper/raw/master/Bundle-Wrapper-0.03.tar.gz
cpanm Bundle-Wrapper-0.03.tar.gz

# 5. Install Tabix
mamba install -y -c bioconda tabix

# 6. Install R packages required by MSG
R -e 'library(devtools);install_github("gabraham/plink2R", subdir="plink2R")'
mamba install -y r-bigmemory r-bigmemory.sri r-bh r-uuid
R -e 'library(devtools);install_github("XingjieShi/TisCoMM")'
mamba install -y r-argparse r-ggplot2 r-plyr r-dplyr r-magrittr r-reshape2 r-data.table r-gtools r-rcolorbrewer r-extrafont r-gridextra r-rmarkdown r-knitr r-formatr r-readr
R -e 'library(devtools);install_github("crotoc/myutils", dep=FALSE)'

# Activate the environment
conda activate $ENV
```

---

### Running the Example

```bash
# Clone scripts
git clone https://github.com/sallymeeyan/MTSG.git
cd MTSG

# Download MSG.sif and data directory
wget https://www.dropbox.com/scl/fo/vly3z0mxawa5x0v9k4elh/ADUtksnvW0Iyo71twsQDNGc?rlkey=vi69osfby7ipnjru9wva1tte3 -O MSG.zip
unzip MSG.zip

# Activate Singularity environment
conda activate /path_to_env/msg

# Generate X matrix for ENSG00000000457
Rscript scripts/runRmd.R --rmd scripts/perGene.process.x.rmd --gene ENSG00000000457 --dir_data data/ --dir_out test/xmatrix/ --output ENSG00000000457 --eval TRUE --ext html

# Generate .cov file
Rscript scripts/generate_db_and_cov.R --input test/xmatrix/ENSG00000000457.select.rsid.vcf.ac.x_all --dir_cov test/cov/ --ref_ld_chr data/LDREF/1000G.EUR.

# Generate Y matrix
Rscript scripts/runRmd.R --rmd scripts/perGene.process.y.rmd --gene ENSG00000000457 --dir_data data/ --dir_out test/ymatrix --output ENSG00000000457 --eval TRUE --ext html

# Run MTSG
Rscript scripts/gtex_comp_MSG_ACAT_GBJ_120522.R --x test/xmatrix/ENSG00000000457.select.rsid.vcf.ac.x_all --y test/ymatrix/ENSG00000000457.all.final.matrix.decomp --model_training --save_model --cov test/cov/ENSG00000000457.select.rsid.vcf.ac.cov.RData --sumstats data/sumstats/clozukscz.sumstats --dir_out test/MTSG/ --verbose TRUE
```

---

## Type I Error Evaluation

Type I error evaluation scripts are located in the `simulation/` folder.
See `simulation/runType1Err.sh` for details.

---

